src/bert_layers/layers.py

```diff
@@ -292,6 +292,32 @@ def forward(
         return attn_out + self.mlp(self.mlp_norm(attn_out))
 
 
+class FlexBertUnpadParallelPreNormLayer(FlexBertLayerBase):
+    """Composes the FlexBERT parallel attention and MLP blocks into a single layer using pre-normalization."""
+
+    def __init__(self, config: FlexBertConfig):
+        super().__init__()
+        self.norm = get_norm_layer(config)
+        self.attn = get_attention_layer(config)
+        self.mlp = get_mlp_layer(config)
+
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        cu_seqlens: torch.Tensor,
+        max_seqlen: int,
+        indices: Optional[torch.Tensor] = None,
+        attn_mask: Optional[torch.Tensor] = None,
+    ) -> torch.Tensor:
+        """Forward pass for a BERT layer, including both attention and MLP.
+
+        Args:
+            hidden_states: (total_nnz, dim)
+            attn_mask: None or (batch, max_seqlen)
+        """
+        attn_out, intermediate_ff = self.attn(self.norm(hidden_states), cu_seqlens, max_seqlen, indices, attn_mask)
+        return hidden_states + attn_out + self.mlp(intermediate_ff)
```

Comment:
I think this implementation won't dispatch the attention and mlp in parallel because the mlp layer might be blocked until the attention calculation is finished, since both `attn_out` and `intermediate_ff` are required before the mlplayer can start processing `intermediate_ff`.